### PR TITLE
testing another approval model

### PR DESCRIPTION
**Pull request by ruseinov**
@ethanfrey FYI

### CODE REVIEW

File: x/nft/approvals.go [hunk 1]
```diff
@@ -83,15 +132,20 @@ func (a Approvals) Exists() bool {
 
 type ApprovalOperations struct {
 	nft       Owned
-	approvals *[]*Approval
+	approvals *[]*ActionApprovals
 }
 
-func NewApprovalOperations(nft Owned, approvals *[]*Approval) *ApprovalOperations {
+func NewApprovalOperations(nft Owned, approvals *[]*ActionApprovals) *ApprovalOperations {
```
[the remark this answers is not in the document]
Comment (ruseinov) [reply x/nft/approvals.go]: @ethanfrey it's not autogen, I just mimicked the previous approach. This is to be able to modify the slice, including it's length.

**Comment by ruseinov** [comment]
@ethanfrey so what do you think about this approach in general? If you are saying we only have a max of 20 elements in a slice - that might be an overkill. We can easily convert a 20-member slice to a map as well.

I think this approach would really shine on a big number of approvals.
There is one more optimisation to be made in that case, and that is having a mapping of address -> action -> index to allow easy access to any approval for a given address.

File: x/nft/approvals.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ethanfrey) [reply x/nft/approvals.go]: Okay, thanks for the clarification

**Comment by ethanfrey** [comment]
In general I agree that maps are nicer with lots of elements. 

One point to keep in mind, is currently, the NFT is loaded from disk on every transaction it is used in. And the cost of *building* the map is likely the cost of traversing it 2 or 3 times. 

So, if we regularly need to search through all the approvals to locate some element more than say 3 times per execution of a transaction, then the map begins to shine, especially with lots of approvals. However, I think both the case of many approvals, as well as searching many times in one tx execution are not expected. 

Please prove me wrong, especially on the second point (about needing to traversing many times in one tx), so I see where the need is. As is, in the cases I would expect, this would actually be slower and some quick `find` logic would be sufficient.

Also, if you have ideas on how to efficiently hold the parsed data in memory in the non-serialized form over many tx/blocks, I would be interested. This would allow many more in-memory optimizations.

**Comment by ethanfrey** [comment]
**However** what I think would be a very nice solution would be to get the best of both worlds.

If it is possible to store the data in a deterministically sorted array on disk, yet get O(log(N)) lookups for matching items, that would give us quick lookups with many approvals, yet not require the processing overhead every event.

I remember https://godoc.org/sort#Search does a binary search given you can make a function that returns true/false over a cut-off.

We just need a way to store the approvals that match the searches. For example, with a struct like:

```
type Approval struct {
  action string
  address []byte
  options interface{}
}
```

We could define a sort function that does (action, address) ASC, and always keep the array sorted with this (adding/removing an item to a sorted array is only O(N) not O(N log N)). 

For search, we can define a function return true if the sort function shows "equal or greater". Then the search will return the first possible hit, and we check if it is the match, or if it is not a match. If it is not a match, then the `(action, address)` combo doesn't exist in the array.

On the other hand, we may want a sort function like (address, action) ASC, which would allow the same equals check, but also allow us to quickly get all actions for a given address (rather than the all addresses for a given action query as the first approach). This is the same issue in how to next maps or build composite indexes in databases, and I trust you to find the more useful ordering.

I just wanted to propose this technique as a way to get the efficient map look-ups you desire without the costly transformation from the deterministic serialized format.

**Comment by ruseinov** [comment]
I will look into it, however looping over an array to build a map from it does not seem to be that costly, especially given we have limited number of actions.  Correct me if I’m wrong. And it allows a decent readable interface to work with approvals and add new logic. Let me prototype a solution you had in mind for us to see of we want to go that way and also think how we would store things in-memory as an exercise.

**Comment by ruseinov** [comment]
We are only looping once with the current structure

**Comment by ruseinov** [comment]
Well, the problem with these is that we actually need to be able to get stuff by action and by address, which makes things a bit more complex in terms of Has and Get, Insert is easy

**Comment by ruseinov** [comment]
ah, and don’t forget expired:)

**Pull request merged**